massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,642 @@
+            && public_key.verify_signature(&hash_2, &signature_2).is_ok())
+    }
+}
+
+/// Create a new Denunciation from 2 SecureShareEndorsement
+impl TryFrom<(&SecureShareEndorsement, &SecureShareEndorsement)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from(
+        (s_e1, s_e2): (&SecureShareEndorsement, &SecureShareEndorsement),
+    ) -> Result<Self, Self::Error> {
+        // Cannot use the same endorsement twice
+        if s_e1 == s_e2 {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // In order to create a Denunciation, there should be the same
+        // slot, index & public key
+        if s_e1.content.slot != s_e2.content.slot
+            || s_e1.content.index != s_e2.content.index
+            || s_e1.content_creator_pub_key != s_e2.content_creator_pub_key
+        {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // Check sig of s_e1 with s_e1.public_key, s_e1.slot, s_e1.index
+        let s_e1_hash_content = EndorsementDenunciation::compute_content_hash(&s_e1.content)?;
+        let s_e1_hash = EndorsementDenunciation::compute_hash_for_sig_verif(
+            &s_e1.content_creator_pub_key,
+            &s_e1.content.slot,
+            &s_e1.content.index,
+            &s_e1_hash_content,
+        )?;
+        // Check sig of s_e2 but with s_e1.public_key, s_e1.slot, s_e1.index
+        let s_e2_hash_content = EndorsementDenunciation::compute_content_hash(&s_e2.content)?;
+        let s_e2_hash = EndorsementDenunciation::compute_hash_for_sig_verif(
+            &s_e1.content_creator_pub_key,
+            &s_e1.content.slot,
+            &s_e1.content.index,
+            &s_e2_hash_content,
+        )?;
+
+        s_e1.content_creator_pub_key
+            .verify_signature(&s_e1_hash, &s_e1.signature)?;
+        s_e1.content_creator_pub_key
+            .verify_signature(&s_e2_hash, &s_e2.signature)?;
+
+        Ok(Denunciation::Endorsement(EndorsementDenunciation {
+            public_key: s_e1.content_creator_pub_key,
+            slot: s_e1.content.slot,
+            index: s_e1.content.index,
+            signature_1: s_e1.signature,
+            signature_2: s_e2.signature,
+            hash_1: s_e1_hash_content,
+            hash_2: s_e2_hash_content,
+        }))
+    }
+}
+
+/// Create a new Denunciation from 2 SecureHeader
+impl TryFrom<(&SecuredHeader, &SecuredHeader)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from((s_bh1, s_bh2): (&SecuredHeader, &SecuredHeader)) -> Result<Self, Self::Error> {
+        // Cannot use the same block header twice
+        // In order to create a Denunciation, there should be the same slot, index & public key
+        if s_bh1.content.slot != s_bh2.content.slot
+            || s_bh1.content_creator_pub_key != s_bh2.content_creator_pub_key
+            || s_bh1.serialized_data == s_bh2.serialized_data
```

Comment:
I guess it's only for tests because this comparison is heavy

Reply:
Add a TODO so this will be done in a follow up